### PR TITLE
Support java UDAF in FE

### PR DESCRIPTION
**Pull request by dirtysalt**
ref: #2474 

The UDAF is also a class. Here is an example

```
class MyUDAF {
   public static class State { // a internal static public class, which has method
       long serializeLength() {}
   }

   State create();
    void destroy(State);
    void update(State, <input args>);
    void serialize(State, java.nio.ByteBuffer)
    void merge(State, java.nio.ByteBuffer);
    RETURN_TYPE finalize(State);
}
```

And the way to create UDAF is like

```
CREATE AGGREGATE  FUNCTION my_add4(INT, INT) RETURNS INT PROPERTIES (
"symbol" = "com.dirlt.SimpleUDF",
"type" = "StarrocksJar",
"file" = "http://localhost:41006/work/myudf.jar"
);
```

```
mysql> show full functions;
+------------------+-------------+---------------+-------------------+------------------------------------------------------------------------------------------------------------------------------------------+
| Signature        | Return Type | Function Type | Intermediate Type | Properties                                                                                                                               |
+------------------+-------------+---------------+-------------------+------------------------------------------------------------------------------------------------------------------------------------------+
| my_add2(INT,INT) | INT         | Scalar        | NULL              | {"symbol":"com.dirlt.SimpleUDF","file":"http://localhost:41006/work/myudf.jar","type":"SRJAR","md5":"4177b7cf98be8df47aa9483b93600846"}  |
| my_add5(INT,INT) | INT         | Aggregate     | NULL              | {"symbol":"com.dirlt.SimpleUDAF","file":"http://localhost:41006/work/myudf.jar","type":"SRJAR","md5":"4177b7cf98be8df47aa9483b93600846"} |
+------------------+-------------+---------------+-------------------+------------------------------------------------------------------------------------------------------------------------------------------+
```

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/analysis/CreateFunctionStmt.java [hunk 1]
```diff
@@ -180,14 +295,33 @@ private void analyzeUdfClassInStarrocksJar() throws AnalysisException {
...
         } catch (MalformedURLException e) {
-            throw new AnalysisException("failed to load object_file: " + objectFile);
+            throw new AnalysisException("Failed to load object_file: " + objectFile);
         } catch (ClassNotFoundException e) {
-            throw new AnalysisException("class '" + class_name + "' not found in object_file :" + objectFile);
+            throw new AnalysisException("Class '" + class_name + "' not found in object_file :" + objectFile);
+        } finally {
+            if (classLoader != null) {
+                try {
+                    classLoader.close();
```
Comment (dirtysalt) [inline fe/fe-core/src/main/java/com/starrocks/analysis/CreateFunctionStmt.java:320]: Better to call `close` here?

Comment (dirtysalt) [inline fe/fe-core/src/main/java/com/starrocks/analysis/CreateFunctionStmt.java:320]: Better to call `close` here?

File: fe/fe-core/src/main/java/com/starrocks/analysis/CreateFunctionStmt.java [hunk 2]
```diff
@@ -80,20 +86,129 @@
...
+                throw new AnalysisException(
+                        String.format("UDF class '%s' method '%s' parameter %s[%s] type does not match %s",
+                                clazz.getCanonicalName(), method.getName(), pname, ptype.getCanonicalName(),
+                                cls.getCanonicalName()));
+            }
+        }
+    }
+
+    private UDFInternalClass udfClass;
+    private UDFInternalClass udfStateClass;
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/analysis/CreateFunctionStmt.java:211]:
```suggestion
    private UDFInternalClass udafClass;
```

File: fe/fe-core/src/main/java/com/starrocks/analysis/CreateFunctionStmt.java [hunk 3]
```diff
@@ -180,14 +295,33 @@ private void analyzeUdfClassInStarrocksJar() throws AnalysisException {
             throw new AnalysisException("No '" + SYMBOL_KEY + "' in properties");
         }
 
+        URLClassLoader classLoader = null;
         try {
             URL[] urls = {new URL("jar:" + objectFile + "!/")};
-            URLClassLoader cl = URLClassLoader.newInstance(urls);
-            udfClass = cl.loadClass(class_name);
+            classLoader = URLClassLoader.newInstance(urls);
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/analysis/CreateFunctionStmt.java:301]: Can be use try (URLClassLoader classLoader = URLClassLoader.newInstance(urls))

File: fe/fe-core/src/main/java/com/starrocks/analysis/CreateFunctionStmt.java [hunk 4]
```diff
@@ -215,93 +349,115 @@ private void computeObjectChecksum() throws IOException, NoSuchAlgorithmExceptio
...
-        }
-        String updateFnSymbol = properties.get(UPDATE_KEY);
-        if (updateFnSymbol == null) {
-            throw new AnalysisException("No 'update_fn' in properties");
-        }
-        String mergeFnSymbol = properties.get(MERGE_KEY);
-        if (mergeFnSymbol == null) {
-            throw new AnalysisException("No 'merge_fn' in properties");
+    private void checkStarrocksJarUdfClass() throws AnalysisException {
+        {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/analysis/CreateFunctionStmt.java:353]: Duplicate `{`

File: fe/fe-core/src/main/java/com/starrocks/analysis/CreateFunctionStmt.java [hunk 5]
```diff
@@ -215,93 +349,115 @@ private void computeObjectChecksum() throws IOException, NoSuchAlgorithmExceptio
...
         }
-        ScalarType scalarType = (ScalarType) type;
-        Class cls = PrimitiveTypeToJavaClassType.get(scalarType.getPrimitiveType());
-        if (cls == null) {
-            throw new AnalysisException("UDF does not support type: " + scalarType);
-        }
-        if (!cls.equals(ptype)) {
-            throw new AnalysisException(String.format("UDF %s[%s] type does not match %s", pname,
-                    ptype.getCanonicalName(), cls.getCanonicalName()));
+        {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/analysis/CreateFunctionStmt.java:385]: Duplicate `{`

File: fe/fe-core/src/main/java/com/starrocks/analysis/CreateFunctionStmt.java [hunk 6]
```diff
@@ -215,93 +349,115 @@ private void computeObjectChecksum() throws IOException, NoSuchAlgorithmExceptio
...
-                throw new AnalysisException(String.format("UDF '%s' parameter count does not match", name));
-            }
-            for (int i = 0; i < method.getParameters().length; i++) {
-                Parameter p = method.getParameters()[i];
-                checkStarrocksJarUdfType(argsDef.getArgTypes()[i], p.getType(), p.getName());
-            }
-        } else {
-            checked = false;
+    private void checkStarrocksJarUdafClass() throws AnalysisException {
+        {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/analysis/CreateFunctionStmt.java:395]: Don't need define multi code block, reuse method is OK, like:
```
Method method;
method = udfClass.getMethod(CREATE_METHOD_NAME, true);
......
......

method = udfClass.getMethod(DESTROY_METHOD_NAME, true);
```

File: fe/fe-core/src/main/java/com/starrocks/analysis/CreateFunctionStmt.java (hunk 2 above)
Comment (dirtysalt) [reply fe/fe-core/src/main/java/com/starrocks/analysis/CreateFunctionStmt.java:211]: `udfStateClass` is internal state class.

File: fe/fe-core/src/main/java/com/starrocks/analysis/CreateFunctionStmt.java (hunk 6 above)
Comment (dirtysalt) [reply fe/fe-core/src/main/java/com/starrocks/analysis/CreateFunctionStmt.java:395]: Yes. I want to split to multi-blocks to emphasize checking of each function.

**Review by stdpain — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Pull request merged**